### DISCUSSION TITLE
defaultValues doesnt work for <select>

### DISCUSSION THREAD

**Discussion by jdvsch** [react-hook-form/react-hook-form#8106] [Q&A] [answered]
Hello,

I am new to react and react-hook-form, so I hope I am writing this in the right place. In the following code that I give you, the default values are correctly displayed in all the fields except in the <select>:

`import React from "react";
import { useForm } from "react-hook-form";

export default function ModalEditarId(props) {
    const { register, handleSubmit } = useForm({
      defaultValues: props.modalEditarId
    });
    const onSubmit = data => console.log(data);

  return (
    <div className="modal fade" id="staticBackdrop" data-bs-backdrop="static" data-bs-keyboard="false" tabIndex="-1" aria-labelledby="staticBackdropLabel" aria-hidden="true">
        <div className="modal-dialog">
            <div className="modal-content">
                <div className="modal-header">
                    <h5 className="modal-title" id="staticBackdropLabel">Editar ID</h5>
                    <button type="button" className="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
                </div>

                <form onSubmit={handleSubmit(onSubmit)}>
                    <div className="modal-body">
                        <div className="row">
                            <div className="col-md-4">
                                <label htmlFor="state" className="col-form-label">State</label>
                                <select {...register("state")} id="state" className="form-select" aria-label="Default select example">
                                    <option defaultValue></option>
                                    <option value="ACTIVE">In use</option>
                                    <option value="INACTIVE">In disuse</option>
                                </select>
                            </div>
                            <div className="col-md-4">
                                <label htmlFor="contratype" className="col-form-label">Contratype</label>
                                <input {...register("contratype")} type="number" id="contratype" className="form-control" />
                            </div>
                            <div className="col-md-4">
                                <label htmlFor="color" className="col-form-label">Color</label>
                                <input {...register("color")} type="text" id="color" className="form-control" />  
                            </div>                
                        </div>
                    </div>
                    <div className="modal-footer">
                        <button type="button" className="btn btn-secondary" data-bs-dismiss="modal">Close</button>
                        <button type="submit" className="btn btn-primary">Submit</button>
                    </div>
                </form> 
            </div>
        </div>
    </div>
  )
}`

It is only displayed correctly if, for example, I write it with either of the following two options:
`<select {...register("state")} defaultValue="ACTIVE" id="state" className="form-select" aria-label="Default select example">`
or
`const { register, handleSubmit } = useForm({
       state: "ACTIVE"
     });`

**Comment by Moshyfawn** [comment]
Please, consider at least formatting your code instead of dumping it for us to dig through.

Better yet, please, consider creating a minimal reproduction codesandbox before we can start looking into your use-case.

**Reply by Moshyfawn** [reply]
[Here's a working codesanbox example](https://codesandbox.io/s/magical-shannon-c35ehv?file=%2Fsrc%2FApp.js). Please, consider tweaking it to match your use-case where you get the issue.

**Reply by Moshyfawn** [reply]
*I've updated the codesanbox to JavaScript instead of TypeScript as you've mentioned that you're new to React

**Reply by jdvsch** [reply]
> *I've updated the codesanbox to JavaScript instead of TypeScript as you've mentioned that you're new to React
Hi Moshyfawn,

excellent idea... I just shared the codesandbox

**Comment by Moshyfawn** [comment, marked as answer]
The reason you're experiencing the issue is because the default form values are cached on the first component render.
[From RHF docs on `useForm`](https://react-hook-form.com/api/useform):
> defaultValues are cached on the first render within the custom hook. If you want to [reset](https://react-hook-form.com/api/useform/reset) the defaultValues, you should use the reset api.

Your `Modal` component is mounted when your app first rendered and isn't unmounted even when it's hidden with the CSS `display: none;` rule.

You can
- **either** unmount your `Modal` component with JS. For example, conditionality render it like:
```jsx
return (
  {!!modalEditarId && <Modal modalEditarId={modalEditarId} />}
)
```
- **or** reset your form's values via [the `reset` form method](https://react-hook-form.com/api/useform/reset) as a side-effect on `modalEditarId` change like
```js
const { reset } = useForm()

useEffect(() => {
  if (modalEditarId) {
    reset(modalEditarId)
  }
}, [reset, modalEditarId])
```

**Reply by jdvsch** [reply]
> The reason you're experiencing the issue is because the default form values are cached on the first component render. [From RHF docs on `useForm`](https://react-hook-form.com/api/useform):
> 
> > defaultValues are cached on the first render within the custom hook. If you want to [reset](https://react-hook-form.com/api/useform/reset) the defaultValues, you should use the reset api.
> 
> Your `Modal` component is mounted when your app first rendered and isn't unmounted even when it's hidden with the CSS `display: none;` rule.
> 
> You can
> 
> * **either** unmount your `Modal` component with JS. For example, conditionality render it like:
> 
> ```js
> return (
>   {!!modalEditarId && <Modal modalEditarId={modalEditarId} />}
> )
> ```
> 
> * **or** reset your form's values via [the `reset` form method](https://react-hook-form.com/api/useform/reset) as a side-effect on `modalEditarId` change like
> 
> ```js
> const { reset } = useForm()
> 
> useEffect(() => {
>   if (modalEditarId) {
>     reset(modalEditarId)
>   }
> }, [reset, modalEditarId])
> ```

I really appreciate you taking the time to review the code and help me understand my mistake. I also appreciate that you included the change that had to be made in the code, since as I said before, I am just learning about this topic and I still have a hard time understanding certain topics.

**Reply by Moshyfawn** [reply]
For sure, no problem!
